crates/ty_ide/src/semantic_tokens.rs

```diff
@@ -0,0 +1,2608 @@
+
+    pub const fn empty() -> Self {
+        Self(0)
+    }
+
+    pub const fn contains(self, other: Self) -> bool {
+        (self.0 & other.0) == other.0
+    }
+
+    #[must_use]
+    pub const fn union(self, other: Self) -> Self {
+        Self(self.0 | other.0)
+    }
+
+    /// Returns all supported token modifiers for LSP capabilities.
+    pub fn all() -> Vec<&'static str> {
+        vec!["definition", "readonly", "async"]
+    }
+
+    /// Convert to LSP modifier indices for encoding
+    pub fn to_lsp_indices(self) -> Vec<u32> {
+        let mut indices = Vec::new();
+        if self.contains(Self::DEFINITION) {
+            indices.push(0);
+        }
+        if self.contains(Self::READONLY) {
+            indices.push(1);
+        }
+        if self.contains(Self::ASYNC) {
+            indices.push(2);
+        }
+        indices
+    }
+}
+
+/// A semantic token with its position and classification.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SemanticToken {
+    pub range: TextRange,
+    pub token_type: SemanticTokenType,
+    pub modifiers: SemanticTokenModifier,
+}
+
+impl Ranged for SemanticToken {
+    fn range(&self) -> TextRange {
+        self.range
+    }
+}
+
+/// The result of semantic tokenization.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SemanticTokens {
+    tokens: Vec<SemanticToken>,
+}
+
+impl SemanticTokens {
+    /// Create a new `SemanticTokens` instance.
+    pub fn new(tokens: Vec<SemanticToken>) -> Self {
+        Self { tokens }
+    }
+}
+
+impl Deref for SemanticTokens {
+    type Target = [SemanticToken];
+
+    fn deref(&self) -> &Self::Target {
+        &self.tokens
+    }
+}
+
+/// Generates semantic tokens for a Python file within the specified range.
+/// Pass None to get tokens for the entire file.
+pub fn semantic_tokens(db: &dyn Db, file: File, range: Option<TextRange>) -> SemanticTokens {
+    let parsed = parsed_module(db, file).load(db);
+    let semantic_model = SemanticModel::new(db, file);
+
+    let mut visitor = SemanticTokenVisitor::new(db, &semantic_model, range);
+    visitor.visit_body(parsed.suite());
+
+    SemanticTokens::new(visitor.tokens)
+}
+
+/// AST visitor that collects semantic tokens.
+struct SemanticTokenVisitor<'db> {
+    #[allow(dead_code)]
+    db: &'db dyn Db,
+    #[allow(dead_code)]
+    semantic_model: &'db SemanticModel<'db>,
+    tokens: Vec<SemanticToken>,
+    in_class_scope: bool,
+    in_type_annotation: bool,
+    range_filter: Option<TextRange>,
+}
+
+impl<'db> SemanticTokenVisitor<'db> {
+    fn new(
+        db: &'db dyn Db,
+        semantic_model: &'db SemanticModel<'db>,
+        range_filter: Option<TextRange>,
+    ) -> Self {
+        Self {
+            db,
+            semantic_model,
+            tokens: Vec::new(),
+            in_class_scope: false,
+            in_type_annotation: false,
+            range_filter,
+        }
+    }
+
+    fn add_token(
+        &mut self,
+        range: TextRange,
```

Comment:
Nit: You could change this to take an `impl Ranged` so that you don't need to call `.range` when e.g. adding a token for `Name` node
```suggestion
        ranged: impl Ranged,
```